storagestate/storagestate.go

```diff
@@ -0,0 +1,730 @@
+		return aoserrors.Wrap(err)
+	}
+
+	return nil
+}
+
+func (storageState *StorageState) checkChecksumAndSendUpdateRequest(
+	stateFilePath string, instanceIdent aostypes.InstanceIdent,
+) (err error) {
+	state, ok := storageState.statesMap[instanceIdent]
+	if !ok {
+		return ErrNotFound
+	}
+
+	_, checksum, err := getFileDataChecksum(stateFilePath)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if !bytes.Equal(checksum, state.checksum) {
+		if err = storageState.pushStateRequestMessage(instanceIdent, false); err != nil {
+			return aoserrors.Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+func (storageState *StorageState) setupStateWatching(instancePath, stateFilePath string, params SetupParams) error {
+	if err := createStateFileIfNotExist(stateFilePath, params.UID, params.GID); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if err := storageState.startStateWatching(instancePath, stateFilePath, params); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	return nil
+}
+
+func (storageState *StorageState) startStateWatching(
+	instancePath string, stateFilePath string, params SetupParams,
+) (err error) {
+	if err = storageState.watcher.Add(stateFilePath); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	storageState.statesMap[params.InstanceIdent] = &stateParams{
+		instancePath:       instancePath,
+		stateFilePath:      stateFilePath,
+		quota:              params.StateQuota,
+		changeTimerChannel: make(chan bool, 1),
+	}
+
+	return nil
+}
+
+func (storageState *StorageState) stopStateWatching(instanceIdent aostypes.InstanceIdent) (err error) {
+	if state, ok := storageState.statesMap[instanceIdent]; ok {
+		if err = storageState.watcher.Remove(state.stateFilePath); err != nil {
+			return aoserrors.Wrap(err)
+		}
+
+		if state.changeTimer != nil {
+			if state.changeTimer.Stop() {
+				state.changeTimerChannel <- false
+			}
+		}
+
+		delete(storageState.statesMap, instanceIdent)
+	}
+
+	return nil
+}
+
+func (storageState *StorageState) pushNewStateMessage(
+	instanceIdent aostypes.InstanceIdent, checksum, stateData string,
+) (err error) {
+	if err := storageState.messageSender.SendInstanceNewState(cloudprotocol.NewState{
+		InstanceIdent: instanceIdent,
+		Checksum:      checksum,
+		State:         stateData,
+	}); err != nil && !errors.Is(err, amqphandler.ErrNotConnected) {
+		return aoserrors.Wrap(err)
+	}
+
+	return nil
+}
+
+func (storageState *StorageState) pushStateRequestMessage(
+	instanceIdent aostypes.InstanceIdent, defaultState bool,
+) (err error) {
+	if err := storageState.messageSender.SendInstanceStateRequest(cloudprotocol.StateRequest{
+		InstanceIdent: instanceIdent,
+		Default:       defaultState,
+	}); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	return nil
+}
+
+func (storageState *StorageState) processWatcher() {
+	for {
+		select {
+		case event, ok := <-storageState.watcher.Events:
+			if !ok {
+				return
+			}
+
+			storageState.Lock()
```

Comment:
Could you move all locked stuff to a separate function.?

Reply:
Done